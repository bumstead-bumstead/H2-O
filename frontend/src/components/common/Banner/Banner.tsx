import type { HTMLAttributes, PropsWithChildren } from 'react';
import styled from '@emotion/styled';
import { Flex, Typography } from '@/components/common';

interface Props extends HTMLAttributes<HTMLDivElement> {
  title?: string;
  subTitle?: string;
  description?: string;
  isTitleColorWhite?: boolean;
  backgroundColor?: string;
}

function Banner({
  children,
  title,
  subTitle,
  description,
  isTitleColorWhite = false,
  backgroundColor,
  ...restProps
}: PropsWithChildren<Props>) {
  return (
    <BannerContainer role='banner' backgroundColor={backgroundColor} {...restProps}>
      <Flex justifyContent='space-between' width='100%' height='100%'>
        {title && (
          <TitleContainer flexDirection='column' gap={4} isTitleColorWhite={isTitleColorWhite}>
            <h3>{subTitle}</h3>
            <h2>{title}</h2>
            {description && <Description font='TextKRRegular14'>{description}</Description>}
          </TitleContainer>
        )}
        {children}
      </Flex>
    </BannerContainer>
  );
}

export default Banner;

const BannerContainer = styled.div<Pick<Props, 'backgroundColor'>>`
  position: relative;
  width: 100%;
  height: 360px;
  background: ${({ backgroundColor }) => backgroundColor};
  box-shadow: 0px 0px 8px 0px rgba(131, 133, 136, 0.2);

  & > div {
    max-width: 1024px;
    margin: 0 auto;
  }
`;

const TitleContainer = styled(Flex)<Pick<Props, 'isTitleColorWhite'>>`
  position: fixed;
  top: 72px;
  width: 400px;

  & > h3 {
    ${({ theme }) => theme.typography.TextKRRegular14}
    color: ${({ isTitleColorWhite, theme }) => (isTitleColorWhite ? 'white' : theme.colors.gray900)};
  }

  & > h2 {
    ${({ theme }) => theme.typography.HeadKRBold32}
    color: ${({ isTitleColorWhite, theme }) => (isTitleColorWhite ? 'white' : theme.colors.primary700)};
  }
`;

const Description = styled(Typography)`
  width: 100%;
  word-break: keep-all;
  margin-bottom: 24px;
`;
