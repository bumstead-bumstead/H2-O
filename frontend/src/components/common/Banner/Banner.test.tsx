import Banner from './Banner';
import { render, screen } from '@/tests/test-util';

describe('Banner 컴포넌트 테스트', () => {
  it('배너가 정상적으로 렌더링된다.', () => {
    render(<Banner />);

    expect(screen.getByRole('banner')).toBeInTheDocument();
  });

  it('title props를 전달하면 텍스트가 정상적으로 표시된다.', () => {
    render(<Banner title='배너' />);

    expect(screen.getByText('배너')).toBeInTheDocument();
  });

  it('subTitle props를 전달하면 텍스트가 정상적으로 표시된다.', () => {
    render(<Banner title='배너' subTitle='서브타이틀' />);

    expect(screen.getByText('서브타이틀')).toBeInTheDocument();
  });

  it('description props를 전달하면 텍스트가 정상적으로 표시된다.', () => {
    render(<Banner title='배너' description='설명' />);

    expect(screen.getByText('설명')).toBeInTheDocument();
  });

  it('isTitleColorWhite props를 전달하면 텍스트가 흰색으로 표시된다.', () => {
    render(<Banner title='배너' isTitleColorWhite />);

    const text = screen.getByText('배너');
    const style = getComputedStyle(text);

    expect(style.color).toBe('rgb(255, 255, 255)');
  });

  it('backgroundColor props를 전달하면 배경색이 정상적으로 표시된다.', () => {
    render(<Banner backgroundColor='black' />);

    const banner = screen.getByRole('banner');
    const style = getComputedStyle(banner);

    expect(style.backgroundColor).toBe('rgb(0, 0, 0)');
  });

  it('배너 스냅샷 테스트', () => {
    render(<Banner title='타이틀' subTitle='서브타이틀' />);

    expect(screen.getByRole('banner')).toMatchSnapshot();
  });
});
