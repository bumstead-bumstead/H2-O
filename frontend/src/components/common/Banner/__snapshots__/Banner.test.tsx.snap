// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Banner 컴포넌트 테스트 > 배너 스냅샷 테스트 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
  height: 360px;
  box-shadow: 0px 0px 8px 0px rgba(131, 133, 136, 0.2);
}

.emotion-0>div {
  max-width: 1024px;
  margin: 0 auto;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
  height: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4px;
  position: fixed;
  top: 72px;
  width: 400px;
}

.emotion-2>h3 {
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 22px;
  letter-spacing: -0.03em;
  color: #212122;
}

.emotion-2>h2 {
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 32px;
  font-weight: 700;
  line-height: 36px;
  letter-spacing: -0.03em;
  color: #00224A;
}

<div
  class="emotion-0"
  role="banner"
>
  <div
    class="emotion-1"
    height="100%"
    width="100%"
  >
    <div
      class="emotion-2"
    >
      <h3>
        서브타이틀
      </h3>
      <h2>
        타이틀
      </h2>
    </div>
  </div>
</div>
`;
